### PR TITLE
fix: fall back to API prompt when opening workflow for API-dispatched jobs

### PR DESCRIPTION
**Pull request by guill**
*PR Created by the Glary-Bot Agent*

---

## Summary

- "Open as workflow in new tab" now works for API-dispatched workflows by falling back to loading the prompt data when `extra_pnginfo.workflow` is not available
- "Export workflow" also falls back to exporting the API-format prompt when the full workflow isn't available

## Problem

API-dispatched workflows (submitted via `/api/prompt`) don't populate `extra_data.extra_pnginfo.workflow` because API clients typically don't send `extra_data`. The "Open as workflow in new tab" and "Export workflow" menu actions only looked in `extra_pnginfo.workflow`, silently failing for API jobs.

## Solution

When `extra_pnginfo.workflow` is missing, the code now falls back to extracting the API-format prompt from `workflow.prompt` (which is always stored for all jobs). The prompt data is loaded via `app.loadApiJson()`, which creates nodes, wires connections, and auto-arranges the layout.

## Changes

| File | Change |
|------|--------|
| `jobTypes.ts` | Extended `zWorkflowContainer` to parse `prompt` field |
| `fetchJobs.ts` | Added `extractPrompt()` to extract API-format prompt from job detail |
| `jobOutputCache.ts` | Added `getJobPrompt()` with LRU caching |
| `useJobMenu.ts` | `openJobWorkflow` falls back to `loadApiJson`; `exportJobWorkflow` falls back to prompt export |
| `getWorkflowFromHistory.test.ts` | Added 6 tests for `extractPrompt` covering happy path and edge cases |

## Verification

Manually tested end-to-end:
1. Started ComfyUI backend + frontend dev server
2. Submitted API workflow via `POST /api/prompt` (no `extra_data`)
3. Confirmed job detail has `prompt` but no `extra_pnginfo.workflow`
4. Clicked "Open as workflow in new tab" → nodes loaded with connections and auto-arranged layout

## Screenshots

Context menu showing Open as workflow in new tab for an API-dispatched job

API workflow loaded successfully with SolidMask, MaskToImage, and SaveImage nodes auto-arranged

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-10929-fix-fall-back-to-API-prompt-when-opening-workflow-for-API-dispatched-jobs-33b6d73d36508186bf6beb43f8ba6e45) by [Unito](https://www.unito.io)

### CODE REVIEW

File: src/platform/remote/comfyui/jobs/fetchJobs.ts [hunk 1]
```diff
@@ -159,3 +162,32 @@ export async function extractWorkflow(
...
+  const isValid = Object.values(prompt).every((node) => {
+    if (!node || typeof node !== 'object' || Array.isArray(node)) return false
+    const record = node as Record<string, unknown>
+    return (
+      typeof record.class_type === 'string' &&
+      record.inputs !== null &&
+      typeof record.inputs === 'object' &&
+      !Array.isArray(record.inputs)
+    )
+  })
```
Comment (guill) [inline src/platform/remote/comfyui/jobs/fetchJobs.ts]: Is there anything specific that this check is protecting against? If so, we should document it in a comment. If not, I would lean towards removing the check unless there's a strong argument that something that *isn't* a prompt might be at this field.

**Comment by christian-byrne** [comment]
## Draft Review — Architectural Feedback

**Verdict: 🟡 Needs course correction**

### What's looking good
- Right diagnosis: API-dispatched jobs really do lack `extra_data.extra_pnginfo.workflow`, and `prompt` is the correct fallback source.
- Cache layer extension is clean — `getJobPrompt` reuses `getJobDetail`, no extra round-trips.
- Test coverage for `extractPrompt` covers the happy path and key edge cases.

### Architectural concerns

1. **`openJobWorkflow` fallback breaks the "new tab" semantic.** The workflow branch uses `workflowStore.createTemporary` + `workflowService.openWorkflow` (opens a tab). The prompt branch calls `app.loadApiJson`, which `rootGraph.clear()`s the **current** workflow in place. The two branches end in observably different states for the user. Consider routing the API path through the temporary-workflow / `openWorkflow` flow as well — e.g., load the prompt into a temp workflow and open it the same way, or convert the prompt to a workflow JSON before the tab-open call. As-is, users will lose unsaved work in the active tab when this fallback fires.

2. **`exportJobWorkflow` silently changes file format.** When falling back to `getJobPrompt`, the exported `Job <id>.json` is an API prompt, not a workflow — the two formats aren't interchangeable on re-import. Either (a) differentiate the filename (e.g., `Job <id>.api.json`), (b) wrap the prompt in a minimal workflow envelope, or (c) prompt the user that only the API prompt is available.

3. **`zWorkflowContainer.prompt` is too loose.** Schema declares `z.record(z.string(), z.unknown())` then `extractPrompt` casts the result `as ComfyApiWorkflow`. `zComfyApiWorkflow` already exists in `workflowSchema.ts` — using it (or `safeParse`-ing it inside `extractPrompt`) would catch malformed prompts at the boundary instead of failing later in `loadApiJson`.

### Suggested direction

Unify the two branches in `openJobWorkflow` so both end up calling `workflowService.openWorkflow` on a temporary workflow, and keep the export format consistent (or signal the difference). Tightening the prompt schema validation is a small change but removes the unsafe cast.

---
*This is an early-stage review focused on direction — detailed line-by-line feedback will come when the PR is marked ready for review.*

**Comment by yoobi** [comment]
Any chance this is getting merged ?
